### PR TITLE
SDK-632 -- Disable Tracking option does not persist throughout multiple sessions.

### PR DESCRIPTION
**Pull request by apeterson-branch**
## Reference
SDK-632 -- Disable Tracking option does not persist throughout multiple sessions.

## Description
We were not saving the Tracking State to shared preferences.   

## Testing Instructions
1. Disable tracking
2. Send an open request and verify it's unsuccessful
3. Close the app
4. Open it again
5. Send another Open Request

Expected Results:  Verify the request was *not* sent, nor should any other requests be sent until Enable Tracking is sent.

## Risk Assessment [`HIGH`]
Lots and lots of changes happened to make this work.

- [X] I, the PR creator, have tested — integration, unit, or otherwise — this code.

## Reviewer Checklist (To be checked off by the reviewer only)

- [ ] JIRA Ticket is referenced in PR title.
- Correctness & Style
    - [ ] Conforms to [Style Guides](https://google.github.io/styleguide/cppguide.html)
    - [ ] Mission critical pieces are documented in code and out of code as needed.
- [ ] Unit Tests reviewed and test issue sufficiently.
- [ ] Functionality was reviewed in QA independently by another engineer on the team.

### CODE REVIEW

File: BranchSDK/src/BranchIO/PropertyManager.h [hunk 1]
```diff
@@ -107,13 +110,48 @@ class BRANCHIO_DLL_EXPORT PropertyManager : protected JSONObject, public virtual
...
+     * @param value Key value
+     */
+    void saveString(const char *path, const char *key, const std::string &value);
+
+    /**
+     * Load a string property from storage.
+     * @param path Path from which this key will be loaded
+     * @param key Key to load
+     * @param defValue bool value from storage, or default value if not found.
+     * @return true if the value was found
```
Comment (bklastaitis-branch) [inline BranchSDK/src/BranchIO/PropertyManager.h]: I believe the current implementation would not return true if the value that was found was `false`

File: BranchSDK/src/BranchIO/SessionInfo.h [hunk 2]
```diff
@@ -47,13 +47,13 @@ class BRANCHIO_DLL_EXPORT SessionInfo : public PropertyManager {
     virtual bool hasSessionId() const;
 
  private:
-    static constexpr const char *const StoragePrefix = "session";
+    static constexpr const char *const SessionStorage = "session";
```
Comment (bklastaitis-branch) [inline BranchSDK/src/BranchIO/SessionInfo.h]: variable that starts with a capital letter

File: BranchSDK/src/BranchIO/SessionInfo.h [hunk 3]
```diff
@@ -47,13 +47,13 @@ class BRANCHIO_DLL_EXPORT SessionInfo : public PropertyManager {
...
 
-    static std::string getStoragePath(const char *key);
+//    static std::string getStoragePath(const char *key);
     virtual SessionInfo& doAddProperty(const char *name, const std::string &value);
 
-    void load(const char* key);
-    void save(const char *key, const std::string &value);
+//    void load(const char* key);
+//    void save(const char *key, const std::string &value);
```
Comment (bklastaitis-branch) [inline BranchSDK/src/BranchIO/SessionInfo.h]: Remove dead code?

File: BranchSDK/src/BranchIO/AdvertiserInfo.h [hunk 4]
```diff
@@ -82,6 +82,9 @@ class BRANCHIO_DLL_EXPORT AdvertiserInfo : public PropertyManager {
     bool trackingDisabled;
     bool trackingLimited;
 
+    static constexpr const char *const AdvertiserStorage = "advertiser";
```
Comment (bklastaitis-branch) [inline BranchSDK/src/BranchIO/AdvertiserInfo.h]: variable that starts with a capital letter

File: BranchSDK/src/BranchIO/Util/WindowsStorage.cpp [hunk 5]
```diff
@@ -132,6 +132,31 @@ WindowsStorage::set(const std::string& key, const std::string& value, Scope scop
     return *this;
 }
 
+bool
+WindowsStorage::getBoolean(const std::string& key, bool& value, Scope scope) const {
+    string registryKey, registryPath;
```
Comment (bklastaitis-branch) [inline BranchSDK/src/BranchIO/Util/WindowsStorage.cpp]: Inconsistency. Elsewhere strings are declared as `std::string`

File: BranchSDK/src/BranchIO/Util/WindowsStorage.cpp [hunk 6]
```diff
@@ -132,6 +132,31 @@ WindowsStorage::set(const std::string& key, const std::string& value, Scope scop
...
+    bool validKey(getRegistryKeyAndPath(scope, key, registryKey, registryPath));
+    assert(validKey);
```
Comment (bklastaitis-branch) [inline BranchSDK/src/BranchIO/Util/WindowsStorage.cpp]: This isn't a test class, so assert looks out of place. Also, you already check for the validity of the key below (i.e. `if (!has(key, scope)) return false;`)

File: BranchSDK/test/StorageTest.cpp [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+#include <gtest/gtest.h>
+
+#include <BranchIO/Util/Storage.h>
+
+using namespace BranchIO;
+using namespace std;
+
+class StorageTest : public ::testing::Test
+{
+    virtual void SetUp() {
```
Comment (bklastaitis-branch) [inline BranchSDK/test/StorageTest.cpp]: method name starting in with capital letter

File: BranchSDK/test/StorageTest.cpp [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        Storage::instance().setDefaultScope(Storage::User);
+        Storage::instance().clear(Storage::User);
+    }
+
+    virtual void TearDown() {
```
Comment (bklastaitis-branch) [inline BranchSDK/test/StorageTest.cpp]: method name starting in with capital letter

File: BranchSDK/test/StorageTest.cpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+
+    bool result = false;
+    bool found = Storage::instance().getBoolean(_testBooleanKey, result);
+
+    ASSERT_TRUE(found);
+    ASSERT_EQ(_testBooleanValue, result);
+}
+
+// Test loading values that are not there
+TEST_F(StorageTest, TestNotFound) {
```
Comment (bklastaitis-branch) [inline BranchSDK/test/StorageTest.cpp]: `TestNotFound` looks like another variable starting with capital letter.

File: BranchSDK/src/BranchIO/SessionInfo.h (hunk 2 above)
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/SessionInfo.h]: Fixed.

File: BranchSDK/src/BranchIO/SessionInfo.h (hunk 3 above)
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/SessionInfo.h]: Fixed

File: BranchSDK/src/BranchIO/AdvertiserInfo.h (hunk 4 above)
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/AdvertiserInfo.h]: Fixed

File: BranchSDK/test/StorageTest.cpp (hunk 7 above)
Comment (apeterson-branch) [reply BranchSDK/test/StorageTest.cpp]: SetUp and TearDown signatures are inherited from google test.

File: BranchSDK/test/StorageTest.cpp (hunk 8 above)
Comment (apeterson-branch) [reply BranchSDK/test/StorageTest.cpp]: SetUp and TearDown signatures are inherited from google test.

File: BranchSDK/src/BranchIO/Util/WindowsStorage.cpp (hunk 6 above)
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/Util/WindowsStorage.cpp]: The original implementation of storage has a ton of `assert`s sprinkled in.

I created a follow-up ticket `SDK-638` to resolve these.

File: BranchSDK/src/BranchIO/Util/WindowsStorage.cpp (hunk 5 above)
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/Util/WindowsStorage.cpp]: There is a lot of inconsistency around the std namespace.  I filed `SDK-639` to follow up on this.

File: BranchSDK/src/BranchIO/PropertyManager.h (hunk 1 above)
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/PropertyManager.h]: Reworked how storage can use default variables.   It was really not useful to have a `has()` method to check if a key exists, and also return a boolean if the variable was loaded from storage.   I updated the method signature to allow for a defaultValue instead, when the key was not found.

File: BranchSDK/src/BranchIO/PropertyManager.cpp [hunk 10]
```diff
@@ -81,6 +81,49 @@ PropertyManager::clear() {
...
+PropertyManager::saveString(const char *path, const char *key, const std::string &value) {
+    Storage::instance().setString(getStoragePath(path, key), value);
+}
+
+bool
+PropertyManager::loadBoolean(const char *path, const char* key, bool defValue) {
+    bool value = defValue;
+    bool result = Storage::instance().getBoolean(getStoragePath(path, key), value);
+    if (result) {
+        addProperty(key, value);
```
[the remark this answers is not in the document]
Comment (apeterson-branch) [reply BranchSDK/src/BranchIO/PropertyManager.cpp]: This was reworked.

File: BranchSDK/src/BranchIO/Util/WindowsStorage.cpp (hunk 5 above)
Comment (bklastaitis-branch) [reply BranchSDK/src/BranchIO/Util/WindowsStorage.cpp]: 👍

File: BranchSDK/src/BranchIO/Util/WindowsStorage.cpp (hunk 6 above)
Comment (bklastaitis-branch) [reply BranchSDK/src/BranchIO/Util/WindowsStorage.cpp]: 👍

File: BranchSDK/test/StorageTest.cpp (hunk 9 above)
Comment (bklastaitis-branch) [reply BranchSDK/test/StorageTest.cpp]: I'm not familiar with C++ unit testing, so gave it a second look and it seems like that's the standard as per Google C++ Testing Framework. Marking as resolved.

File: BranchSDK/src/BranchIO/PropertyManager.h (hunk 1 above)
Comment (bklastaitis-branch) [reply BranchSDK/src/BranchIO/PropertyManager.h]: 💯

**Comment by bklastaitis-branch** [comment]
I resolved stuff you already addressed, let me know when the Windows storage is fixed and I'll go through this one more time. Also, seems like there are merging conflicts.

**Comment by apeterson-branch** [comment]
I added 3 new unit tests to investigate @bklastaitis-branch latest comment regarding setting the default scope.   Unit tests pass on both Unix and Windows.

If the scope is set to `Default`, the code will call `getDefaultScope()` which is currently set at `User`.

```
    if (scope == Default) scope = getDefaultScope();
```

**Pull request merged**